### PR TITLE
Add DistributeGeneric pass

### PR DESCRIPTION
**Pull request by j2kun**
First part of https://github.com/google/heir/issues/170

Next step is to figure out how to handle distributing through ops that have nested regions.

### CODE REVIEW

File: lib/Dialect/Secret/Transforms/DistributeGeneric.cpp [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+      : OpRewritePattern<GenericOp>(context, /*benefit=*/1) {}
+
+  LogicalResult matchAndRewrite(GenericOp op,
+                                PatternRewriter &rewriter) const override {
+    bool hasUnusedOps = false;
+    Block *body = op.getBody();
+    for (BlockArgument arg : body->getArguments()) {
+      if (arg.use_empty()) {
+        hasUnusedOps = true;
+        rewriter.updateRootInPlace(op, [&]() {
```
Comment (mr0re1) [inline lib/Dialect/Secret/Transforms/DistributeGeneric.cpp:82]: Is it safe to remove by index? Would next iteration `arg.getArgNumber()` be decremented "magically" ?

**Review by mr0re1 — APPROVED** [review]

File: lib/Dialect/Secret/Transforms/DistributeGeneric.cpp (hunk 1 above)
Comment (j2kun) [reply lib/Dialect/Secret/Transforms/DistributeGeneric.cpp:82]: It looks like the implementation does indeed update argNumber for all arguments: https://github.com/llvm/llvm-project/blob/9c5b38b49fa2a1a81dac78faa0a8432eff495615/mlir/lib/IR/Block.cpp#L192

File: lib/Dialect/Secret/Transforms/DistributeGeneric.cpp [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+    }
+
+    Operation &firstOp = body->getOperations().front();
+    // The inputs to the op are generic op's block arguments (cleartext values),
+    // and they need to change to be the corresponding generic op's normal
+    // operands (maybe secret values).
+    SmallVector<Value> newInputs;
+    // The indices of the new inputs in the original block argument list
+    SmallVector<unsigned> newInputIndices;
+    for (Value val : firstOp.getOperands()) {
```
Comment (asraa) [inline lib/Dialect/Secret/Transforms/DistributeGeneric.cpp:146]: do you need to check that `firstOp`'s operands don't use the secret inputs? (In the example case - that the firstOp does not use `clear_value`.)

Comment (j2kun) [reply lib/Dialect/Secret/Transforms/DistributeGeneric.cpp:146]: IMO this is something that should be enforced by the verifier: all ops within a secret.generic can only use as operands the values in the args of the containing basic block, and those args have to correspond to the cleartext versions of the secret.generic's operands in the same order. Added to https://github.com/google/heir/issues/108

File: lib/Dialect/Secret/Transforms/DistributeGeneric.cpp (hunk 1 above)
Comment (j2kun) [reply lib/Dialect/Secret/Transforms/DistributeGeneric.cpp:82]: Update, you were right :) It passes tests, but hits an assertion which apparently only runs in debug mode internally

Updating to

```
    for (int i = 0; i < body->getArguments().size(); ++i) {
      BlockArgument arg = body->getArguments()[i];
      if (arg.use_empty()) { 
        hasUnusedOps = true;  
        rewriter.updateRootInPlace(op, [&]() {
          body->eraseArgument(i);
          op.getOperation()->eraseOperand(i);
        });
        // Ensure the next iteration uses the right arg number
        --i;
      }
    }
```

File: lib/Dialect/Secret/Transforms/DistributeGeneric.cpp (hunk 2 above)
Comment (asraa) [reply lib/Dialect/Secret/Transforms/DistributeGeneric.cpp:150]: Oh, that part I understand - but my question is that it seems like from this description
```
// Split a secret.generic containing multiple ops into its underlying op if no
// secret values are included in its input
```

that this intends to peel apart the first op only when the op doesnt use the block arguments.

Now reading the code, it seems like maybe that comment is incorrect, and actually this is just breaking apart the secret.generic ops?

Comment (j2kun) [reply lib/Dialect/Secret/Transforms/DistributeGeneric.cpp:150]: Ah, yeah I think I had originally intended that, and then decided to split that into its own separate rewrite rule for simplicity. Will update the comment.

**Review by asraa — APPROVED** [review]

**Comment by j2kun** [comment]
Sigh, due to copybara import issues we had to merge this in a special commit https://github.com/google/heir/commit/dda03f5d4f8bea3597bf44e26b7994f36f8c2cb5

Closing

**Pull request closed without merging**